### ISSUE TITLE
[Proposal] std::any Layer::getProperty(string);

### ISSUE THREAD

**Issue by zhoonit** [nnstreamer/nntrainer#1297]
For now, we have some (theoretical) properties frequently queried from other objects.
For example, flatten, trainable, supportInPlace, requireLabels are those kind of stuff.

# As is
To support this, we are making one api each to query this. But this can be problemetic, if we are considering the list is keep growing. and If we add a new API, we don't want to update the version. (or think about tizen capi)

# To Be

Move this to a layer property and make it available in LayerImpl level. When querying, we use std::any.

```cpp
class ActivationLayer: LayerImpl {
   ActivationLayer() : LayerImpl({"trainable=false", "require_label=false"}): {
     /** ... */
   }
}; 

TEST(query_trainable_option1) {
 auto layer = ActivationLayer();
 std::any is_trainable = layer.getProperties("trainable"); // if trainable is NA or empty, this will be empty
 EXPECT_NE(false, is_trainable.empty());
 bool &is_trainable_ = std::any_cast<std::reference_wrapper<bool>>(is_trainable).get();
 EXPECT_EQ(false, is_trainable_);
 is_trainable_ = true; // this will change original properties
}

TEST(query_trainable_option2) {
 auto layer = ActivationLayer();
 std::any is_trainable = layer.getProperties("trainable");
 EXPECT_EQ(false, std::any_cast<bool>(is_trainable).get()); // in this option, we cannot modify variable
 is_trainable = true; // this won't change the original properties
}
```

**Comment by kparichay** [comment]
With this change, developers will have to use c++17 for their custom layers. Although our code is c++17, we did not force custom developers to also c++17 but with std::any it will.
How about starting with a simplified signature of returning bool/int (it currently satisfied minimal requirement), and add std::any with for developers who have c++17?

**Comment by zhoonit** [comment]
We could consider void * (which c api will anyway)
I'll start with bool for now.

**Comment by kparichay** [comment]
Yes, we can provide `void *`, which internally depends on `std::any`, and `std::any` version can be directly used if the developer uses c++17.
